client/me/purchases/cancel-purchase/index.jsx

```diff
@@ -18,41 +18,69 @@ import Main from 'components/main';
 import paths from '../paths';
 import titles from 'me/purchases/titles';
 import { getName, isCancelable } from 'lib/purchases';
-import { getPurchase, goToManagePurchase, isDataLoading, recordPageView } from '../utils';
+import { getPurchase, getSelectedSite, goToManagePurchase, recordPageView } from 'me/purchases/utils';
 
 const CancelPurchase = React.createClass( {
 	propTypes: {
 		selectedPurchase: React.PropTypes.object.isRequired,
-		selectedSite: React.PropTypes.object
+		selectedSite: React.PropTypes.object.isRequired
+	},
+
+	getInitialState() {
+		return {
+			isRedirecting: false
+		};
 	},
 
 	componentWillMount() {
-		recordPageView( 'cancel_purchase', this.props );
+		if ( this.isDataValid( this.props ) ) {
+			recordPageView( 'cancel_purchase', this.props );
+		} else {
+			this.redirect( this.props );
+		}
 	},
 
 	componentWillReceiveProps( nextProps ) {
-		recordPageView( 'cancel_purchase', this.props, nextProps );
+		if ( this.isDataValid( nextProps ) ) {
+			recordPageView( 'cancel_purchase', this.props, nextProps );
+		} else {
+			this.redirect( nextProps );
+		}
 	},
 
-	componentDidMount() {
-		this.ensurePageCanLoad();
+	isDataValid( props ) {
+		const purchase = getPurchase( props ),
+			selectedSite = getSelectedSite( props );
+
+		return ( selectedSite && purchase && isCancelable( purchase ) );
 	},
 
-	componentDidUpdate() {
-		this.ensurePageCanLoad();
+	redirect( props ) {
+		if ( this.state.isRedirecting ) {
```

Comment:
Instead of adding state to an otherwise stateless component, can we just check if the data is invalid here?

```js
if ( ! this.isDataValid() ) {
  return null;
}
```